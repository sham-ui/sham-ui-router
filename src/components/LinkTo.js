import { inject, options } from 'sham-ui';
import { onclick, ref } from 'sham-ui-directives';
import LinkToComponentTemplate from './LinkTo.sht';

export default class LinkToComponent extends LinkToComponentTemplate {
    constructor( options ) {
        super( {
            ...options,
            directives: {
                onclick,
                ref,
                ...options.directives
            }
        } );
    }

    @inject router;
    @inject( 'sham-ui:store' ) componentStore;

    @options text = '';
    @options useActiveClass = false;
    @options activeClass = 'active';
    @options className = '';

    @options get params() {
        return {};
    }

    generateURL( path, params ) {
        return this.lastGeneratedURL = this.router.generate( path, params );
    }

    // eslint-disable-next-line no-unused-vars
    isActive( path, params ) {
        return this.lastGeneratedURL === this.router.lastRouteResolved().url;
    }

    generateClassName( className, useActiveClass, activeClass, path, params ) {
        return [
            className,
            useActiveClass && this.isActive( path, params ) ? activeClass : ''
        ].join( ' ' ).trim();
    }

    _registryType() {
        const activeLinkType = ACTIVE_PAGE_LINK_TYPE;
        const storeHasType = this.componentStore.byType.has( activeLinkType );
        if ( this.options.useActiveClass ) {
            if ( !storeHasType ) {
                this.componentStore.byType.set( activeLinkType, new Set() );
            }
            this.componentStore.byType.get( activeLinkType ).add( this );
        } else if ( storeHasType ) {
            this.componentStore.byType.get( activeLinkType ).delete( this );
        }
    }

    update() {
        super.update( ...arguments );
        this._registryType();
    }

    click( e ) {
        e.preventDefault();
        this.router.navigate( this.aNode.getAttribute( 'href' ) );
    }
}
