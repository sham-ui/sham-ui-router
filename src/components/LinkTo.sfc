<template>
    <a
            href={{this.generateURL( path, params )}}
            class={{this.generateClassName( className, useActiveClass, activeClass, path, params )}}
            :onclick={{::this.click}}
            :ref="aNode"
    >
        {{text}}
        {% defblock %}
    </a>
</template>

<script>
    import { inject, options } from 'sham-ui';
    import { onclick, ref } from 'sham-ui-directives';

    export default class LinkToComponent extends Template {
        constructor( options ) {
            super( {
                ...options,
                directives: {
                    onclick,
                    ref,
                    ...options.directives
                }
            } );
        }

        @inject router;
        @inject( 'sham-ui:store' ) componentStore;

        @options text = '';
        @options useActiveClass = false;
        @options activeClass = 'active';
        @options className = '';

        @options get params() {
            return {};
        }

        generateURL( path, params ) {
            return this.lastGeneratedURL = this.router.generate( path, params );
        }

        // eslint-disable-next-line no-unused-vars
        isActive( path, params ) {
            return this.lastGeneratedURL === this.router.lastRouteResolved().url;
        }

        generateClassName( className, useActiveClass, activeClass, path, params ) {
            return [
                className,
                useActiveClass && this.isActive( path, params ) ? activeClass : ''
            ].join( ' ' ).trim();
        }

        _registryType() {
            const activeLinkType = ACTIVE_PAGE_LINK_TYPE;
            const storeHasType = this.componentStore.byType.has( activeLinkType );
            if ( this.options.useActiveClass ) {
                if ( !storeHasType ) {
                    this.componentStore.byType.set( activeLinkType, new Set() );
                }
                this.componentStore.byType.get( activeLinkType ).add( this );
            } else if ( storeHasType ) {
                this.componentStore.byType.get( activeLinkType ).delete( this );
            }
        }

        update() {
            super.update( ...arguments );
            this._registryType();
        }

        click( e ) {
            e.preventDefault();
            this.router.navigate( this.aNode.getAttribute( 'href' ) );
        }
    }
</script>