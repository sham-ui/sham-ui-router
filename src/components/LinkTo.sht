<a
    href={{this.generateURL( path, params )}}
    class={{this.generateClassName( className, useActiveClass, activeClass, path, params )}}
    :onclick={{::this.click}}
    :ref="aNode"
>
    {{text}}
    {% defblock %}
</a>