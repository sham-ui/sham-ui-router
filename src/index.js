import { DI, inject } from 'sham-ui';
import Navigo from 'navigo';

export default class Router {
    @inject UI = 'sham-ui';

    constructor( root = null, useHash = false, hash = '#', autoResolve = true ) {
        DI.bind( 'router', this );
        this.router = new Navigo( root, useHash, hash );
        this.activePageWidget = null;
        this.activePageOptions = null;
        this.activePageLinks = new Set();
        if ( autoResolve ) {
            this.UI.render.on( 'RegistrationComplete', this.resolve.bind( this ) );
        }
    }

    resolve() {
        this.router.resolve();
    }

    bindPage( url, name, pageWidget, widgetOptions ) {
        this.on( url, {
            as: name,
            uses: () => {
                this.activePageWidget = pageWidget;
                this.activePageOptions = widgetOptions;
                this._renderActivatePage();
            }
        } );
        return this;
    }

    _renderActivatePage() {
        const widget = DI.resolve( 'widgets:active-page-container' );
        if ( undefined !== widget ) {
            this.UI.render.ONLY( widget.ID );
            this.activePageLinks.forEach( x => x.update() )
        }
    }

    on() {
        this.router.on.apply( this.router, arguments );
    }

    off() {
        this.router.off.apply( this.router, arguments );
    }

    notFound() {
        this.router.notFound.apply( this.router, arguments );
    }

    navigate() {
        this.router.navigate.apply( this.router, arguments );
    }

    hooks() {
        this.router.hooks.apply( this.router, arguments );
    }

    destroy() {
        this.router.destroy.apply( this.router, arguments )
    }

    link() {
        return this.router.link.apply( this.router, arguments );
    }

    lastRouteResolved() {
        return this.router.lastRouteResolved.apply( this.router, arguments );
    }

    generate() {
        return this.router.generate.apply( this.router, arguments );
    }

    _registerActivePageLink( widget ) {
        this.activePageLinks.add( widget );
    }

    _unregisterActivePageLink( widget ) {
        this.activePageLinks.delete( widget );
    }
}