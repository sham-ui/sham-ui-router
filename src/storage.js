import createStorage from 'sham-ui-data-storage';
/**
 * Data storage for router service
 * @see https://github.com/sham-ui/sham-ui-data-storage
 * @typedef {Object} RouterStorage
 * @property {string} url Current page url
 * @property {string} name Current page name
 * @property {Object} params Current page params
 * @property {Class<Component>} activePageComponent Current page component class
 * @property {Object} activePageOptions Options for current page component
 */

export const { storage, useStorage } = createStorage( {
    url: '',
    name: '',
    params: {},
    activePageComponent: null,
    activePageOptions: null
}, {
    DI: 'router:storage',
    sync: true
} );
