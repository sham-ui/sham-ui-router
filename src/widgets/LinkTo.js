import { inject, options } from 'sham-ui';
import { onclick } from 'sham-ui-directives';
import LinkToWidgetTemplate from './LinkTo.sht';

export default class LinkToWidget extends LinkToWidgetTemplate {
    constructor( selector, id, options ) {
        super( selector, id, {
            ...options,
            directives: {
                onclick,
                ...options.directives
            }
        } );
    }

    @inject router = 'router';

    @options get text() {
        return '';
    }
    @options get params() {
        return {};
    }
    @options get useActiveClass() {
        return false;
    }
    @options get activeClass() {
        return 'active';
    }
    @options get className() {
        return '';
    }

    generateURL( path, params ) {
        return this.lastGeneratedURL = this.router.generate( path, params );
    }

    // eslint-disable-next-line no-unused-vars
    isActive( path, params ) {
        return this.lastGeneratedURL === this.router.lastRouteResolved().url;
    }
    _registryInRouter() {
        if ( this.options.useActiveClass ) {
            this.router._registerActivePageLink( this );
        }
    }
    _removeFromRouter() {
        if ( this.options.useActiveClass ) {
            this.router._unregisterActivePageLink( this );
        }
    }
    get aNode() {
        return this.querySelector( 'a' );
    }
    update() {
        super.update( ...arguments );
        this._registryInRouter();
    }
    destroy() {
        this._removeFromRouter();
    }
    remove() {
        super.remove( ...arguments );
        this._removeFromRouter();
    }
    click( e ) {
        e.preventDefault();
        this.router.navigate( this.aNode.getAttribute( 'href' ) );
    }
}
