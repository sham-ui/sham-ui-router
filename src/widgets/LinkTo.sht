<a
    href="{{this.generateURL( path, params )}}"
    class="{{
        [
            className,
            useActiveClass && this.isActive( path, params ) ? activeClass : ''
        ].join( ' ' ).trim()
    }}"
    :onclick={{this.click.bind( this )}}
>
    {{text}}
</a>