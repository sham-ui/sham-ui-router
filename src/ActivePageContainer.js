import { $ } from 'sham-ui-macro/ref.macro';
import { Component, insert } from 'sham-ui';
import { useStorage } from './storage';

/**
 * Options for LinkTo
 * @typedef {Object} LinkToOptions
 * @property {string} path Name of page for link. Default ''
 * @property {Object} params Params of page for link. Default {}
 * @property {string} text Text for link. Default ''
 * @property {boolean} useActiveClass Use activeClass options for active page link. Default false
 * @property {string} activeClass Class name for active link. Default 'active'
 */

/**
 * Component for link to page
 * @class LinkTo
 * @classdesc
 * @property {LinkToOptions} options
 */

/**
 * Component-container for page
 * @example
 * {% import ActivePageContainer from 'sham-ui-router/active-page-container' %}
 * ...
 * <ActivePageContainer/>
 * ...
 */
function ActivePageContainer() {
    this.ref = null;

    let lastRenderedURL = null;
    let lastRenderedComponent = null;

    this.spots = [ [
        $.routerData,
        routerData => {
            const url = routerData.url;
            const activePageComponent = routerData.activePageComponent;
            if (
                url !== lastRenderedURL ||

                // Component can changed without change url
                activePageComponent !== lastRenderedComponent
            ) {

                // Remove self from parent's children map or child ref.
                if ( this.unbind ) {
                    this.unbind();
                }

                // Remove all nested views.
                let i = this.nested.length;
                while ( i-- ) {
                    this.nested[ i ].remove();
                }
            }
            if ( null === activePageComponent ) {
                return;
            }
            insert(
                this,
                this.container,
                this,
                activePageComponent,
                routerData.activePageOptions,
                this.owner,
                this.blocks
            );
            lastRenderedURL = url;
            lastRenderedComponent = activePageComponent;
        }
    ] ];
}

export default Component( useStorage( $.routerData ), ActivePageContainer );
