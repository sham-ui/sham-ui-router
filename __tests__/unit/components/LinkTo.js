import { createDI } from 'sham-ui';
import { storage } from '../../../src/storage';
import LinkTo from '../../../src/components/LinkTo.sfc';
import hrefto from '../../../src/directives/href-to';
import renderer from 'sham-ui-test-helpers';

it( 'render correctly', () => {
    const DI = createDI();

    const generateMock = jest.fn();
    DI.bind( 'router', {
        storage: storage( DI ),
        generate: generateMock
    } );

    generateMock.mockReturnValueOnce( '/base' );

    const meta = renderer( LinkTo, {
        path: 'base',
        text: 'Base page'
    }, {
        DI,
        directives: {
            'hrefto': hrefto
        }
    } );
    expect( meta.toJSON() ).toMatchSnapshot();
    expect( generateMock.mock.calls ).toHaveLength( 1 );
    expect( generateMock.mock.calls[ 0 ] ).toEqual( [ 'base', {} ] );
} );

it( 'params options', () => {
    const DI = createDI();

    const generateMock = jest.fn();
    DI.bind( 'router', {
        storage: storage( DI ),
        generate: generateMock
    } );

    generateMock.mockReturnValueOnce( '/base/1' );

    const meta = renderer( LinkTo, {
        path: 'base',
        params: {
            id: 1
        }
    }, {
        DI,
        directives: {
            'hrefto': hrefto
        }
    } );
    expect( meta.toJSON() ).toMatchSnapshot();
    expect( generateMock.mock.calls ).toHaveLength( 1 );
    expect( generateMock.mock.calls[ 0 ] ).toEqual( [ 'base', { id: 1 } ] );
} );

it( 'useActiveClass options', () => {
    const DI = createDI();

    const generateMock = jest.fn();

    const routerStorage = storage( DI );

    DI.bind( 'router', {
        storage: routerStorage,
        generate: generateMock
    } );

    generateMock.mockReturnValue( '/base' );
    routerStorage.url = '/base';
    routerStorage.sync();

    const meta = renderer( LinkTo, {
        path: 'base',
        useActiveClass: true
    }, {
        DI,
        directives: {
            'hrefto': hrefto
        }
    } );
    expect( meta.toJSON() ).toMatchSnapshot();
    expect( generateMock.mock.calls ).toHaveLength( 1 );
    expect( generateMock.mock.calls[ 0 ] ).toEqual( [ 'base', {} ] );
} );

it( 'activeClass options', () => {
    const DI = createDI();

    const generateMock = jest.fn();

    const routerStorage = storage( DI );

    DI.bind( 'router', {
        storage: routerStorage,
        generate: generateMock
    } );

    generateMock.mockReturnValueOnce( '/base' );
    routerStorage.url = '/base';
    routerStorage.sync();

    const meta = renderer( LinkTo, {
        path: 'base',
        useActiveClass: true,
        activeClass: 'test-active'
    }, {
        DI,
        directives: {
            'hrefto': hrefto
        }
    } );
    expect( meta.toJSON() ).toMatchSnapshot();
} );

it( 'className option', () => {
    const DI = createDI();

    const generateMock = jest.fn();
    DI.bind( 'router', {
        storage: storage( DI ),
        generate: generateMock
    } );

    generateMock.mockReturnValueOnce( '/base' );

    const meta = renderer( LinkTo, {
        path: 'base',
        text: 'Base page',
        className: 'foo bar'
    }, {
        DI,
        directives: {
            'hrefto': hrefto
        }
    } );
    expect( meta.ctx.container.querySelector( 'a' ).className ).toEqual( 'foo bar' );
    expect( meta.toJSON() ).toMatchSnapshot();
} );

it( 'className option & useActiveClass & activeClass', () => {
    const DI = createDI();

    const generateMock = jest.fn();

    const routerStorage = storage( DI );
    DI.bind( 'router', {
        storage: routerStorage,
        generate: generateMock
    } );

    generateMock.mockReturnValueOnce( '/base' );
    routerStorage.url = '/base';
    routerStorage.sync();

    const meta = renderer( LinkTo, {
        path: 'base',
        useActiveClass: true,
        activeClass: 'test-active',
        className: 'foo'
    }, {
        DI,
        directives: {
            'hrefto': hrefto
        }
    } );
    expect(
        meta.ctx.container.querySelector( 'a' ).className
    ).toEqual( 'foo test-active' );
    expect( meta.toJSON() ).toMatchSnapshot();
} );

it( 'click', () => {
    const DI = createDI();

    const generateMock = jest.fn();
    const navigateMock = jest.fn();

    DI.bind( 'router', {
        storage: storage( DI ),
        generate: generateMock,
        navigate: navigateMock
    } );

    generateMock.mockReturnValueOnce( '/base' );

    const meta = renderer( LinkTo, {
        path: 'base'
    }, {
        DI,
        directives: {
            'hrefto': hrefto
        }
    } );
    meta.ctx.container.querySelector( 'a' ).click();

    expect( meta.toJSON() ).toMatchSnapshot();
    expect( navigateMock.mock.calls ).toHaveLength( 1 );
    expect( navigateMock.mock.calls[ 0 ] ).toEqual( [ '/base' ] );
} );

it( 'remove', () => {
    const DI = createDI();

    const generateMock = jest.fn();
    DI.bind( 'router', {
        storage: storage( DI ),
        generate: generateMock
    } );

    generateMock.mockReturnValueOnce( '/base' );

    const meta = renderer( LinkTo, {
        path: 'base',
        text: 'Base page'
    }, {
        DI,
        directives: {
            'hrefto': hrefto
        }
    } );
    meta.component.remove();
    expect( meta.ctx.container.innerHTML ).toEqual( '' );
} );
