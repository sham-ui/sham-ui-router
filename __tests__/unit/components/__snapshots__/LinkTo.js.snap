// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`activeClass options 1`] = `
Object {
  "Options": Object {
    "activeClass": "test-active",
    "className": "",
    "params": Object {},
    "path": "base",
    "text": "",
    "useActiveClass": true,
  },
  "html": "
<a class=\\"test-active\\" href=\\"/base\\">
  <!--0-->
</a>
",
}
`;

exports[`className option & useActiveClass & activeClass 1`] = `
Object {
  "Options": Object {
    "activeClass": "test-active",
    "className": "foo",
    "params": Object {},
    "path": "base",
    "text": "",
    "useActiveClass": true,
  },
  "html": "
<a class=\\"foo test-active\\" href=\\"/base\\">
  <!--0-->
</a>
",
}
`;

exports[`className option 1`] = `
Object {
  "Options": Object {
    "activeClass": "active",
    "className": "foo bar",
    "params": Object {},
    "path": "base",
    "text": "Base page",
    "useActiveClass": false,
  },
  "html": "
<a class=\\"foo bar\\" href=\\"/base\\">Base page
  <!--0-->
</a>
",
}
`;

exports[`click 1`] = `
Object {
  "Options": Object {
    "activeClass": "active",
    "className": "",
    "params": Object {},
    "path": "base",
    "text": "",
    "useActiveClass": false,
  },
  "html": "
<a class=\\"\\" href=\\"/base\\">
  <!--0-->
</a>
",
}
`;

exports[`params options 1`] = `
Object {
  "Options": Object {
    "activeClass": "active",
    "className": "",
    "params": Object {
      "id": 1,
    },
    "path": "base",
    "text": "",
    "useActiveClass": false,
  },
  "html": "
<a class=\\"\\" href=\\"/base/1\\">
  <!--0-->
</a>
",
}
`;

exports[`render correctly 1`] = `
Object {
  "Options": Object {
    "activeClass": "active",
    "className": "",
    "params": Object {},
    "path": "base",
    "text": "Base page",
    "useActiveClass": false,
  },
  "html": "
<a class=\\"\\" href=\\"/base\\">Base page
  <!--0-->
</a>
",
}
`;

exports[`useActiveClass options 1`] = `
Object {
  "Options": Object {
    "activeClass": "active",
    "className": "",
    "params": Object {},
    "path": "base",
    "text": "",
    "useActiveClass": true,
  },
  "html": "
<a class=\\"active\\" href=\\"/base\\">
  <!--0-->
</a>
",
}
`;
